#17 feat(JWT) login logout feature

uridonget opened the pull request:
# feat(JWT) login & logout feature \#17
## Overview
1. JWT 검증 함수 추가
2. JWT를 통해 User 데이터를 불러오는 함수 추가(login)
3. 유저 쿠키의 JWT를 삭제하는 함수 추가(logout)

<!---- Resolves: #(Isuue Number) -->
## PR Type

- [ ] feat (authenticate_token, login, logout)

## PR Checklist
PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.

uridonget committed:
#17(JWT): JWT 검증 및 갱신하는 함수 추가

authenticate_token 함수 추가
    - 사용자 쿠키의 access_token, refresh_token을 검증
    - 토큰 검증 성공/토큰 갱신 성공/토큰 검증 실패로 분기
앞으로 작성되는 API 요청 로직에 삽입하여 권한 검증에 사용
토큰 갱신 성공 시 프론트엔드에서 추가 처리 필요

uridonget committed:
#17(JWT): JWT를 통해 유저 데이터를 불러오는 함수 추가

login 함수 추가
    - JWT가 유효하면 user_data와 상태 코드 반환
    - JWT가 유효하지 않으면 에러 메시지와 상태 코드 반환
    - access_token 갱신 시 “유효하지 않은 것”으로 간주되어 별도 처리 필요
서버 상태를 변경하지 않으므로 GET 메소드 사용

uridonget committed:
#17(JWT): 쿠키의 JWT를 삭제하는 함수 추가

logout 함수 추가
    - JWT가 유효하면 유저의 쿠키에 저장된 JWT를 삭제하고 로그아웃 성공메시지 반환
    - JWT가 유효하지 않으면 에러 메시지와 상태 코드 반환
서버 상태를 변경하지 않으므로 GET 메소드 사용

uridonget committed:
#17(JWT): 인증 관련 에러 코드 수정

authenticate_token 함수에서 Access Token 갱신 후에도 401 Unauthorized를 반환하도록 수정
    - 기존에는 갱신 성공 시 200을 반환했으나, 토큰을 인증한다는 해당 함수의 목적에 맞게 사용되지 못하였기에 수정하였음

JWT에서 추출한 'user_id'와 동일한 id를 가진 유저가 존재하지 않는 경우 401 Unauthorized를 반환하도록 수정
    - 기존에는 404 Not Found를 반환했으나, 존재하지 않는 사용자를 공격자가 추론할 가능성을 고려하여 수정하였음

allorak333 commented on backend/account/views.py:
```diff
@@ -424,3 +425,125 @@ def local_auth_sign_up(request):
+    @details JWT 검증 함수(authenticate_token)의 결과가 유효하면 유저 데이터를 반환한다.
+    """
+    try:
+        user, token_response = authenticate_token(request)
+        if token_response:
+            return token_response
```
```authenticate_token``` 함수에서 token을 refresh하는 경우, refresh token이 만료된 경우 message를 리턴하기 때문에 로그아웃 시도에 불필요한 4xx 코드를 받는 경우가 생기는 문제가 있어보입니다.

uridonget replied:
확실히... refresh token이 만료된 경우에는 이미 로그아웃이나 다름이 없는데 에러코드를 받는 것이 좀 이상하긴 하네요. 그렇다면 logout 함수는 따로 authenticate_token 함수를 사용하지 않고 최적화된 로직으로 하드코딩해서 수정하도록 하겠습니다!

allorak333 commented on backend/account/views.py:
```diff
@@ -424,3 +425,125 @@ def local_auth_sign_up(request):
+            refresh = RefreshToken(refresh_token)
+            new_access = refresh.access_token
+
+            response = JsonResponse(
+                {"message": "Token refreshed successfully."}, status=401
+            )
+            response.set_cookie(
+                key="access_token",
+                value=str(new_access),
+                httponly=True,
+                secure=True,
+                samesite="Lax",
+                max_age=settings.SIMPLE_JWT["ACCESS_TOKEN_LIFETIME"].total_seconds(),
+                path="/",
+            )
+            return None, response
+        except TokenError as e:
+            return None, JsonResponse(
+                {"error": "Invalid or expired refresh token"}, status=401
+            )
+
+    if not user_id:
+        return None, JsonResponse(
+            {"error": "Invalid token payload: user_id missing"}, status=401
+        )
+    try:
+        user = User.objects.get(id=user_id)
+    except User.DoesNotExist:
+        return None, JsonResponse({"error": "User not found"}, status=401)
+    return user, None
```
token 자체가 유효하지 않은 경우와 token이 refresh되는 경우, refresh token이 만료되어 새로 로그인이 필요한 경우 모두 같은 코드로 되어있습니다. 로그인 구현 시 프론트에서는 페이지 이동 여부를 메세지 내용으로 확인해야하는데 이러면 하드코딩을 해야하는 상황이 우려됩니다!

uridonget replied:
저도 그 부분을 생각해보지 않은 것은 아니지만... 진짜 위 상황에 적절한 에러코드를 찾을 수 없었습니다ㅠ
=> 할당되지 않는 상태코드(452 ~ 499)에 저희만의 상태코드를 지정해서 처리하도록 해보겠습니다!

uridonget replied:
수정된 사항은 API 명세서에 업데이트 해놓겠습니다~

allorak333 commented on backend/account/views.py:
```diff
@@ -424,3 +425,125 @@ def local_auth_sign_up(request):
+        return JsonResponse({"error": str(e)}, status=500)
+
+
+@api_view(["GET"])
+def logout(request):
```
GET 요청은 기본적으로 어떠한 컨텐츠를 받기 위해 쓰이는 method입니다. 또한 GET요청의 경우 브라우저에서는 prefetch라는 기능이 있어 의도하지 않게 로그아웃 되는 현상이 발생할 수 있습니다. 따라서 logout에는 GET보다는 POST혹은 DELETE가 더 적절하다고 생각합니다.

uridonget replied:
GET 요청의 prefetch 기능에 대해서 좀 알아보니 로그아웃 기능은 절대! GET 요청으로 보내면 안될 것 같다는 생각이 드네요. 서버 측의 데이터에 변화가 없다(브라우저의 쿠키를 삭제하는게 끝)는 점에서 GET method로 설정한 것이었는데 확실히 다른 method를 사용하는 것이 더 좋을 것 같습니다. DELETE는 따로 삭제하는 데이터가 (아직은)없으니 POST로 수정하도록 하겠습니다!

uridonget replied:
동일한 이유로 login 함수도 POST로 수정하도록 하겠습니다!

uridonget committed:
#17 Chore(message): Response 리턴값 수정

기존에는 4xx~5xx 상태코드를 가지면 응답에 error 필드를 넣고 2xx~3xx 상태코드를 가지면 message 필드를 넣었는데
이것을 message 필드로 통일하여 프론트에서 작업을 용이하게 할 수 있도록 수정하였음.

uridonget committed:
#17 Chore(response): Response 상태코드 수정

authenticate_token 함수에서 access_token을 refresh 하는 경우에는 상태코드 401이 아니라 452를 반환하도록 수정.

uridonget committed:
#17 Chore(HTTP method): login & logout method 수정

login 함수와 logout 함수가 POST method 요청으로 처리되도록 수정.

uridonget committed:
#17 Fix(logout): logout 로직 변경

기존에는 authenticate_token을 통해 토큰의 유효성을 검사하고나서 토큰을 삭제하는 로직을 실행하였는데
로그아웃 로직에서 토큰의 유효성을 검사할 필요가 없을 것 같다는 생각에 토큰 검증과정을 삭제하였습니다.

allorak333 commented:
LGTM! 👍

allorak333 reviewed (approved):
LGTM! 👍
